File: ops/src/infinicore/ops/paged_caching/paged_caching.cc
#include "infinicore/ops/paged_caching.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(PagedCaching);

PagedCaching::PagedCaching(Tensor k_cache, Tensor v_cache, const Tensor &k, const Tensor &v, const Tensor &slot_mapping) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(k_cache, v_cache, k, v, slot_mapping);
    INFINICORE_GRAPH_OP_DISPATCH(k->device().getType(), k_cache, v_cache, k, v, slot_mapping);
}

void PagedCaching::execute(Tensor k_cache, Tensor v_cache, const Tensor &k, const Tensor &v, const Tensor &slot_mapping) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(PagedCaching, k_cache, v_cache, k, v, slot_mapping);
}

void paged_caching_(Tensor k_cache, Tensor v_cache, const Tensor &k, const Tensor &v, const Tensor &slot_mapping) {
    PagedCaching::execute(k_cache, v_cache, k, v, slot_mapping);
}

} // namespace infinicore::op

File: python/infinicore/ops/paged_caching.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def paged_caching(
    k_cache: Tensor,
    v_cache: Tensor,
    k: Tensor,
    v: Tensor,
    slot_mapping: Tensor,
):
    Tensor(
        _infinicore.paged_caching_(
            k_cache._underlying,
            v_cache._underlying,
            k._underlying,
            v._underlying,
            slot_mapping._underlying,
        )
    )
    return (k_cache, v_cache)
